url2a908001cd_malictus_klang_tgz-pJ8-malictus_klang_test_PrimitiveTestJ8/src/malictus/klang/KlangUtil.java

```diff
@@ -0,0 +1,720 @@
+     */
+    public static void insertIntoFile(File source, File target, long start) throws IOException {
+		RandomAccessFile rafDest = new RandomAccessFile(target, "rw");
+		RandomAccessFile rafSource = new RandomAccessFile(source, "r");
+		byte[] buf = new byte[KlangConstants.BUFFER_SIZE];
+		try {
+			long oldend = rafDest.length();
+			long moveAmt = source.length();
+			rafDest.setLength(rafDest.length() + moveAmt);
+			rafDest.seek(oldend);
+			long curpos;
+			while ((rafDest.getFilePointer() - KlangConstants.BUFFER_SIZE) >= start) {
+				curpos = rafDest.getFilePointer();
+				KlangProgressWindow.progVal = KlangProgressWindow.PROG_MAX - (int)(((float)(curpos - start) / (float)(oldend - start)) * KlangProgressWindow.PROG_MAX);
+				rafDest.seek(curpos - KlangConstants.BUFFER_SIZE);
+				int x = rafDest.read(buf);
+				if (x != buf.length) {
+					throw new IOException(KlangConstants.ERROR_GENERAL_FILE);
+				}
+				rafDest.seek(curpos - KlangConstants.BUFFER_SIZE + moveAmt);
+				rafDest.write(buf);
+				rafDest.seek(curpos - KlangConstants.BUFFER_SIZE);
+			}
+			if (rafDest.getFilePointer() != start) {
+				buf = new byte[(int)(rafDest.getFilePointer() - start)];
+				rafDest.seek(start);
+				int x = rafDest.read(buf);
+				if (x != (rafDest.getFilePointer() - start)) {
+					throw new IOException(KlangConstants.ERROR_GENERAL_FILE);
+				}
+				rafDest.seek(start + moveAmt);
+				rafDest.write(buf);
+			}
+			buf = new byte[KlangConstants.BUFFER_SIZE];
+			rafSource.seek(0);
+			rafDest.seek(start);
+			while ((rafSource.getFilePointer() + KlangConstants.BUFFER_SIZE) <= rafSource.length()) {
+				KlangProgressWindow.progVal = KlangProgressWindow.PROG_MAX - (int)(((float)(rafSource.length() - rafSource.getFilePointer()) / (float)(rafSource.length() - start)) * KlangProgressWindow.PROG_MAX);
+				int x = rafSource.read(buf);
+				if (x != buf.length) {
+					throw new IOException(KlangConstants.ERROR_GENERAL_FILE);
+				}
+				rafDest.write(buf);
+			}
+			if (rafSource.getFilePointer() < rafSource.length()) {
+				buf = new byte[(int)(rafSource.length() - rafSource.getFilePointer())];
+				int x = rafSource.read(buf);
+				if (x != buf.length) {
+					throw new IOException(KlangConstants.ERROR_GENERAL_FILE);
+				}
+				rafDest.write(buf);
+			}
+			rafDest.close();
+			rafSource.close();
+		} catch (IOException err) {
+			rafDest.close();
+			rafSource.close();
+			throw err;
+		} 
+    }
+    
+	/**
+	 * Copy a file.
+	 * @param source the source file to copy
+	 * @param dest the destination file; if file already exists, it will be overwritten
+	 * @throws IOException if read/write error occurs
+	 */
+    public static void copyFile(File source, File dest) throws IOException {
+    	if (dest.getPath().equals(source.getPath())) {
+			throw new IOException(KlangConstants.ERROR_IN_OUT_IDENTICAL);
+		}
+    	if (dest.exists()) {
+			dest.delete();
+		}
+		dest.createNewFile();
+		FileInputStream fin = new FileInputStream(source);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fin*. The resource is referenced at line 531. The resource is closed at lines 534, 537. There are other execution paths that do not contain closure statements, for example, when FileOutputStream constructor throws an exception. Ensure *fin* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).